libsolidity/formal/CHC.cpp

```diff
@@ -164,27 +196,125 @@ bool CHC::visit(IfStatement const& _if)
+}
+
+bool CHC::visit(WhileStatement const& _while)
+{
+	eraseKnowledge();
+	m_context.resetVariables(touchedVariables(_while));
+	return false;
+}
+
+bool CHC::visit(ForStatement const& _for)
+{
+	eraseKnowledge();
+	m_context.resetVariables(touchedVariables(_for));
 	return false;
 }
 
 void CHC::endVisit(FunctionCall const& _funCall)
 {
+	m_functionCallSeen = true;
+
 	solAssert(_funCall.annotation().kind != FunctionCallKind::Unset, "");
 
-	if (_funCall.annotation().kind == FunctionCallKind::FunctionCall)
+	if (_funCall.annotation().kind != FunctionCallKind::FunctionCall)
+	{
+		SMTEncoder::endVisit(_funCall);
+		return;
+	}
+
+	FunctionType const& funType = dynamic_cast<FunctionType const&>(*_funCall.expression().annotation().type);
+	switch (funType.kind())
 	{
-		FunctionType const& funType = dynamic_cast<FunctionType const&>(*_funCall.expression().annotation().type);
-		if (funType.kind() == FunctionType::Kind::Assert)
-			visitAssert(_funCall);
+	case FunctionType::Kind::Assert:
+		visitAssert(_funCall);
+		SMTEncoder::endVisit(_funCall);
+		break;
+	case FunctionType::Kind::Internal:
+	case FunctionType::Kind::External:
+	case FunctionType::Kind::DelegateCall:
+	case FunctionType::Kind::BareCall:
+	case FunctionType::Kind::BareCallCode:
+	case FunctionType::Kind::BareDelegateCall:
+	case FunctionType::Kind::BareStaticCall:
+	case FunctionType::Kind::Creation:
+	case FunctionType::Kind::KECCAK256:
+	case FunctionType::Kind::ECRecover:
+	case FunctionType::Kind::SHA256:
+	case FunctionType::Kind::RIPEMD160:
+	case FunctionType::Kind::BlockHash:
+	case FunctionType::Kind::AddMod:
+	case FunctionType::Kind::MulMod:
+		SMTEncoder::endVisit(_funCall);
+		unknownFunctionCall(_funCall);
+		break;
+	default:
+		SMTEncoder::endVisit(_funCall);
+		break;
 	}
+}
+
+void CHC::visitAssert(FunctionCall const& _funCall)
+{
+	auto const& args = _funCall.arguments();
+	solAssert(args.size() == 1, "");
+	solAssert(args.front()->annotation().type->category() == Type::Category::Bool, "");
 
-	SMTEncoder::endVisit(_funCall);
+	solAssert(!m_path.empty(), "");
+
+	smt::Expression assertNeg = !(m_context.expression(*args.front())->currentValue());
+	smt::Expression assertionError = smt::Expression::implies(
+		m_path.back() && m_context.assertions() && assertNeg,
+		error()
+	);
+	string predicateName = "assert_" + to_string(_funCall.id());
+	m_interface->addRule(assertionError, predicateName + "_to_error");
+
+	m_verificationTargets.push_back(&_funCall);
 }
 
-void CHC::visitAssert(FunctionCall const&)
+void CHC::unknownFunctionCall(FunctionCall const& _funCall)
 {
+	/// Function calls are handled at the moment,
+	/// so always erase knowledge.
+	/// TODO remove when function calls get predicates/blocks.
+	eraseKnowledge();
+
+	FunctionDefinition const* funDef = functionCallToDefinition(_funCall);
+	if (!funDef)
+		return;
+
+	if (funDef->returnParameterList())
+		for (auto const& retParam: funDef->returnParameters())
+			createVariable(*retParam);
+
```

Comment:
The following code is used to create unknown symbolic values for function returns. This is needed for assignments.